### PR TITLE
TMC#95 Sdl built-in functions

### PR DESCRIPTION
**Pull request by rphh**

### CODE REVIEW

File: sdl2promela/builtins.py [hunk 1]
```diff
@@ -24,29 +31,127 @@ def __translate_writeln(
...
+        .build()
+    )
+
+
+def __translate_to_enum(
+    call: sdlmodel.ProcedureCall, parameters: List[promelamodel.Expression]
+):
+    if len(parameters) != 2:
+        raise ValueError(
+            f"to_enum built-in function expected 2 arguments, but received {len(parameters)}."
```
Comment (Lurkerpas) [inline sdl2promela/builtins.py:86]: Why 2? What happens with the second one? Should the subtracted value be 1 or e.g. min?

File: sdl2promela/builtins.py [hunk 2]
```diff
@@ -24,29 +31,127 @@ def __translate_writeln(
...
+        .build()
+    )
+
+
+def __translate_num(
+    call: sdlmodel.ProcedureCall, parameters: List[promelamodel.Expression]
+):
+    if len(parameters) != 1:
+        raise ValueError(
+            f"num built-in function expected 2 arguments, but received {len(parameters)}."
```
Comment (Lurkerpas) [inline sdl2promela/builtins.py:117]: 2?

File: sdl2promela/builtins.py [hunk 3]
```diff
@@ -24,29 +31,127 @@ def __translate_writeln(
...
+        )
+    return parameters[0]
+
+
+def __translate_val(
+    call: sdlmodel.ProcedureCall, parameters: List[promelamodel.Expression]
+):
+    if len(parameters) != 2:
```
Comment (Lurkerpas) [inline sdl2promela/builtins.py:125]: I wonder thether this parameter checking and exception throwing could be extracted into a single method _checkParameterCount(parameters, "function_name", expectedCount).
Also, in some cases they are checked for nulls... in some they are not. Why the inconsistency?

File: sdl2promela/sdl/model.py [hunk 4]
```diff
@@ -907,17 +917,23 @@ def convert(source: ogAST.PrimSequenceOf):
     for elem in source.value:
         elements.append(convert(elem))
 
-    return SequenceOf(elements, source.exprType)
+    sequenceof = SequenceOf(elements, source.exprType)
```
Comment (Lurkerpas) [inline sdl2promela/sdl/model.py:920]: Type assignment cannot be handled in the constructor?

File: sdl2promela/translator.py [hunk 5]
```diff
@@ -473,9 +495,35 @@ def __generate_expression(context: Context, expression: sdlmodel.ProcedureCall):
     # into separate expressions for the calls, assigning results to intermediate values
     # and then using this values in the actual expression
     # TODO
-    raise NotImplementedError(
-        f"Call to {str(expression.name)} cannot be made within an expression"
-    )
+
+    if expression.name.lower() == "present":
+        # A special case when context shall be changed
+        if len(expression.parameters) != 1:
```
Comment (Lurkerpas) [inline sdl2promela/translator.py:501]: THis could also benefit from the parameter check extraction

File: sdl2promela/translator.py [hunk 6]
```diff
@@ -502,11 +550,13 @@ def __generate_procedure_inline(
     builder.withName(__get_procedure_inline_name(context, procedure.name))
     blockBuilder = BlockBuilder(promelamodel.BlockType.BLOCK)
     for localVariable, localVariableTypeObject in procedure.variables.items():
-        localVariableType = resolve_asn1_type(
-            context.sdl_model.types, localVariableTypeObject[0]
-        )
+        # localVariableType = resolve_asn1_type(
```
Comment (Lurkerpas) [inline sdl2promela/translator.py:553]: Why the change and why this is commented out?

File: sdl2promela/translator.py [hunk 7]
```diff
@@ -522,13 +572,15 @@ def __generate_procedure_inline(
                 parameter.name
             )
             builder.withParameter(intermediateParameterName)
-            parameterType = resolve_asn1_type(
-                context.sdl_model.types, parameter.typeObject
-            )
+            # parameterType = resolve_asn1_type(
```
Comment (Lurkerpas) [inline sdl2promela/translator.py:575]: Commented out code

File: sdl2promela/translator.py [hunk 8]
```diff
@@ -624,8 +676,8 @@ def __generate_execute_transition(
     # Generate assignment of signal parameters into variables
     for target_variable_ref in input_block.target_variables:
         variable = context.sdl_model.variables[target_variable_ref.variableName]
-        variableType = resolve_asn1_type(context.sdl_model.types, variable.type)
-        assignInlineName = __get_assign_value_inline_name(variableType)
+        # variableType = resolve_asn1_type(context.sdl_model.types, variable.type)
```
Comment (Lurkerpas) [inline sdl2promela/translator.py:679]: Commented out code

File: sdl2promela/translator.py [hunk 9]
```diff
@@ -991,10 +1043,10 @@ def __generate_statement(
     assert isinstance(transition.parent, sdlmodel.Procedure)
     if procedureReturn.expression is None:
         return None
-    resolvedType = resolve_asn1_type(
-        context.sdl_model.types, transition.parent.returnType
-    )
-    assignInlineName = __get_assign_value_inline_name(resolvedType)
+    # resolvedType = resolve_asn1_type(
```
Comment (Lurkerpas) [inline sdl2promela/translator.py:1046]: Commented out code

File: sdl2promela/translator.py [hunk 10]
```diff
@@ -1066,17 +1118,25 @@ def __generate_assignment(
     right: sdlmodel.Expression,
     left_type: type,
 ) -> List[promelamodel.Statement]:
-    finalType = resolve_asn1_type(context.sdl_model.types, left_type)
+    # finalType = resolve_asn1_type(context.sdl_model.types, left_type)
```
Comment (Lurkerpas) [inline sdl2promela/translator.py:1121]: Commented out code

File: sdl2promela/translator.py [hunk 11]
```diff
@@ -1518,7 +1578,8 @@ def __generate_assignment(
 
     valueType = finalType.Children[right.choice].type
 
-    selection = finalType.CName + "_" + right.choice + "_PRESENT"
+    # selection = finalType.CName + "_" + right.choice + "_PRESENT"
```
Comment (Lurkerpas) [inline sdl2promela/translator.py:1581]: Commented out code

File: sdl2promela/translator.py [hunk 12]
```diff
@@ -1920,8 +1981,10 @@ def __generate_implicit_variable_definition(
     context: Context, variable_name: str, variable_info: sdlmodel.VariableInfo
 ) -> promelamodel.VariableDeclaration:
     mangled_name = __get_implicit_variable_name(context, variable_name)
-    memberType = resolve_asn1_type(context.sdl_model.types, variable_info.type)
-    return VariableDeclarationBuilder(mangled_name, memberType.CName).build()
+    # memberType = resolve_asn1_type(context.sdl_model.types, variable_info.type)
```
Comment (Lurkerpas) [inline sdl2promela/translator.py:1984]: Commented out code

File: sdl2promela/builtins.py [hunk 13]
```diff
@@ -24,29 +31,127 @@ def __translate_writeln(
 
 
 def __translate_length(
-    call: sdlmodel.ProcedureCall, parameters: List[promelamodel.Statement]
-) -> promelamodel.Statement:
-    """
-    Length assigns an integer scalar, so there is no need to pack it into assign_value call
-    """
+    call: sdlmodel.ProcedureCall, parameters: List[promelamodel.Expression]
```
[the remark this answers is not in the document]
Comment (rphh) [reply sdl2promela/builtins.py:34]: The comment is not revelant anymore.

File: sdl2promela/translator.py (hunk 6 above)
Comment (rphh) [reply sdl2promela/translator.py:553]: For the types like `MyBetterInteger ::= MyInteger (0..10)` the generated code was wrong:
```
MyInteger_assign_value(src, dst);
```
After these changes, the correct inline is used to assign values:
```
MyBetterInteger_assign_value(src, dst);
```

File: sdl2promela/translator.py (hunk 12 above)
Comment (rphh) [reply sdl2promela/translator.py]: changed

File: sdl2promela/translator.py (hunk 11 above)
Comment (rphh) [reply sdl2promela/translator.py:1570]: changed

File: sdl2promela/translator.py (hunk 10 above)
Comment (rphh) [reply sdl2promela/translator.py]: changed

File: sdl2promela/translator.py (hunk 8 above)
Comment (rphh) [reply sdl2promela/translator.py]: changed

File: sdl2promela/translator.py (hunk 7 above)
Comment (rphh) [reply sdl2promela/translator.py]: changed

File: sdl2promela/sdl/model.py (hunk 4 above)
Comment (rphh) [reply sdl2promela/sdl/model.py:920]: It was easier to modify (base class and places where the objects are created)

File: sdl2promela/builtins.py (hunk 3 above)
Comment (rphh) [reply sdl2promela/builtins.py]: changed.

File: sdl2promela/builtins.py (hunk 2 above)
Comment (rphh) [reply sdl2promela/builtins.py]: changed

File: sdl2promela/translator.py (hunk 5 above)
Comment (rphh) [reply sdl2promela/translator.py:501]: I prefer to not use private function from builtins.py

File: sdl2promela/builtins.py (hunk 1 above)
Comment (Lurkerpas) [reply sdl2promela/builtins.py]: Please add comment regarding to_enum scope and the fact that choice selectors are always indexed from 0. Please comment why -1.

Comment (Lurkerpas) [reply sdl2promela/builtins.py]: Please comment that this handled the two argument to_enum flavour, where first argument is the choice, and the second, ignored one, is the enum type.

File: sdl2promela/translator.py (hunk 11 above)
Comment (Lurkerpas) [reply sdl2promela/translator.py:1570]: Are you sure?

File: sdl2promela/builtins.py (hunk 1 above)
Comment (rphh) [reply sdl2promela/builtins.py]: added

File: sdl2promela/translator.py (hunk 11 above)
Comment (rphh) [reply sdl2promela/translator.py]: it should be ok now

**Review by Lurkerpas — APPROVED** [review]
